File: java/src/main/java/org/hyperledger/fabric/client/Commit.java
/*
 * Copyright 2021 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package org.hyperledger.fabric.client;

import java.util.function.UnaryOperator;

import io.grpc.CallOptions;

/**
 * Allows access to information about a transaction that is committed to the ledger.
 */
public interface Commit extends Signable {
    /**
     * Get the transaction ID.
     * @return A transaction ID.
     */
    String getTransactionId();

    /**
     * Get the status of the committed transaction. If the transaction has not yet committed, this method blocks until
     * the commit occurs.
     * @return Transaction commit status.
     * @throws CommitStatusException if the gRPC service invocation fails.
     */
    default Status getStatus() throws CommitStatusException {
        return getStatus(GatewayUtils.asCallOptions());
    }

    /**
     * Get the status of the committed transaction. If the transaction has not yet committed, this method blocks until
     * the commit occurs.
     * @param options Function that transforms call options.
     * @return Transaction commit status.
     * @throws CommitStatusException if the gRPC service invocation fails.
     */
    Status getStatus(UnaryOperator<CallOptions> options) throws CommitStatusException;

    /**
     * Get the status of the committed transaction. If the transaction has not yet committed, this method blocks until
     * the commit occurs.
     * @param options Call options.
     * @return Transaction commit status.
     * @throws CommitStatusException if the gRPC service invocation fails.
     * @deprecated Replaced by {@link #getStatus(UnaryOperator)}.
     */
    @Deprecated
    default Status getStatus(CallOption... options) throws CommitStatusException {
        return getStatus(GatewayUtils.asCallOptions(options));
    }
}

File: node/src/commit.ts
/*
 * Copyright 2021 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { CallOptions } from '@grpc/grpc-js';
import { gateway } from '@hyperledger/fabric-protos';
import { GatewayClient } from './client';
import { Signable } from './signable';
import { SigningIdentity } from './signingidentity';
import { Status, StatusCode } from './status';

/**
 * Allows access to information about a transaction that is committed to the ledger.
 */
export interface Commit extends Signable {
    /**
     * Get status of the committed transaction. If the transaction has not yet committed, this method blocks until the
     * commit occurs.
     * @param options - gRPC call options.
     * @throws {@link CommitStatusError}
     * Thrown if the gRPC service invocation fails.
     */
    getStatus(options?: CallOptions): Promise<Status>;

    /**
     * Get the ID of the transaction.
     */
    getTransactionId(): string;
}

export interface CommitImplOptions {
    client: GatewayClient;
    signingIdentity: SigningIdentity;
    transactionId: string;
    signedRequest: gateway.SignedCommitStatusRequest;
}

export class CommitImpl implements Commit {
    readonly #client: GatewayClient;
    readonly #signingIdentity: SigningIdentity;
    readonly #transactionId: string;
    readonly #signedRequest: gateway.SignedCommitStatusRequest;

    constructor(options: Readonly<CommitImplOptions>) {
        this.#client = options.client;
        this.#signingIdentity = options.signingIdentity;
        this.#transactionId = options.transactionId;
        this.#signedRequest = options.signedRequest;
    }

    getBytes(): Uint8Array {
        return this.#signedRequest.serializeBinary();
    }

    getDigest(): Uint8Array {
        const request = this.#signedRequest.getRequest_asU8();
        return this.#signingIdentity.hash(request);
    }

    async getStatus(options?: CallOptions): Promise<Status> {
        await this.#sign();
        const response = await this.#client.commitStatus(this.#signedRequest, options);
        return this.#newStatus(response);
    }

    getTransactionId(): string {
        return this.#transactionId;
    }

    setSignature(signature: Uint8Array): void {
        this.#signedRequest.setSignature(signature);
    }

    async #sign(): Promise<void> {
        if (this.#isSigned()) {
            return;
        }

        const signature = await this.#signingIdentity.sign(this.getDigest());
        this.setSignature(signature);
    }

    #isSigned(): boolean {
        const signatureLength = this.#signedRequest.getSignature().length || 0;
        return signatureLength > 0;
    }

    #newStatus(response: gateway.CommitStatusResponse): Status {
        const code = response.getResult();
        return {
            blockNumber: BigInt(response.getBlockNumber()),
            code,
            successful: code === StatusCode.VALID,
            transactionId: this.#transactionId,
        };
    }
}
